src/core/filesystem.cc

```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
+  Status GetDirectoryFiles(
+      const std::string& path, std::set<std::string>* files) override;
+  Status ReadTextFile(const std::string& path, std::string* contents) override;
+  Status WriteTextFile(
+      const std::string& path, const std::string& contents) override;
+
+ private:
+  google::cloud::StatusOr<gcs::Client> client_;
+  Status ParsePath(
+      const std::string& path, std::string* bucket, std::string* object);
+  Status MetaDataExists(
+      const std::string path, bool* exists,
+      google::cloud::StatusOr<gcs::ObjectMetadata>* metadata);
+};
+
+GCSFileSystem::GCSFileSystem()
+{
+  client_ = gcs::Client::CreateDefaultClient();
+}
+
+Status
+GCSFileSystem::CheckClient()
+{
+  // Need to return error status if GOOGLE_APPLICATION_CREDENTIALS is not set or
+  // valid
+  if (!client_) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "Unable to create GCS client. Check account credentials.");
+  }
+  return Status::Success;
+}
+
+Status
+GCSFileSystem::ParsePath(
+    const std::string& path, std::string* bucket, std::string* object)
+{
+  // Get the bucket name and the object path. Return error if input is malformed
+  size_t bucket_start = path.find("gs://") + 5;
+  size_t bucket_end = path.find("/", bucket_start);
+  *bucket = path.substr(bucket_start, bucket_end - bucket_start);
+  *object = path.substr(bucket_end + 1);
+
+
+  if (bucket->empty()) {
+    return Status(
+        RequestStatusCode::INTERNAL, "No bucket name found in path: " + path);
+  }
+
+  if (object->empty()) {
+    return Status(
+        RequestStatusCode::INTERNAL, "No file name found in path: " + path);
+  }
+
+  return Status::Success;
+}
+
+std::string
+AppendSlash(std::string& name)
+{
+  // Helper function to take care of lack of trailing slashes
+  if (name.empty()) {
+    return "/";
+  }
+
+  if (name.back() != '/') {
+    return (name + "/");
+  }
+  return name;
+}
+
+
+Status
+GCSFileSystem::FileExists(const std::string& path, bool* exists)
+{
+  *exists = false;
+
+  std::string bucket, object;
+  RETURN_IF_ERROR(ParsePath(path, &bucket, &object));
+
+  // Make a request for metadata and check the response
+  google::cloud::StatusOr<gcs::ObjectMetadata> object_metadata =
+      client_->GetObjectMetadata(bucket, object);
+
+  if (object_metadata) {
+    *exists = true;
+  }
+
+  return Status::Success;
+}
+
+Status
+GCSFileSystem::IsDirectory(const std::string& path, bool* is_dir)
+{
+  *is_dir = false;
+  std::string bucket, object_path;
+  RETURN_IF_ERROR(ParsePath(path, &bucket, &object_path));
+
+  // Check if the bucket exists
+  google::cloud::StatusOr<gcs::BucketMetadata> bucket_metadata =
+      client_->GetBucketMetadata(bucket);
+
+  if (!bucket_metadata) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "Could not get MetaData for bucket with name " + bucket);
+  }
+
+  // Check if whether it has children. If at least one child, it is a directory
```

Comment:
So GCS doesn't allow empty directories?

Reply:
It does but the way I've done it here, empty directories will just have themselves as a child. I added a check to remove these is GetDirectoryContents().